pkg/threescale/connectors/backend/cached_backend.go

```diff
@@ -0,0 +1,261 @@
+)
+
+// Cacheable - defines the required behaviour of a Backend cache
+type Cacheable interface {
+	Get(cacheKey string) (CacheValue, bool)
+	Set(cacheKey string, cv CacheValue)
+}
+
+// Limit captures the current state of the rate limit for a particular time period
+type Limit struct {
+	current int
+	max     int
+	// unix timestamp for end period
+	periodEnds int64
+}
+
+// CacheValue which should be stored in cache implementation
+type CacheValue map[string]map[backend.LimitPeriod]*Limit
+
+// CachedBackend provides a pluggable cache enabled 'Backend' implementation
+// Supports reporting metrics for non-cached responses.
+type CachedBackend struct {
+	ReportFn      metrics.ReportMetricsFn
+	cache         Cacheable
+	mutex         sync.RWMutex
+	hierarchyTree hierarchyTree
+}
+
+type LocalCache struct {
+	ds cmap.ConcurrentMap
+}
+
+// map to represent parent --> children metric relationship
+type metricParentToChildren map[string][]string
+
+// hierarchyTree - represents the relationship between service metrics and their children
+type hierarchyTree map[string]metricParentToChildren
+
+// NewCachedBackend returns a pointer to a CachedBackend with a default LocalCache if no custom implementation has been provided
+func NewCachedBackend(cache Cacheable, rfn metrics.ReportMetricsFn) *CachedBackend {
+	if cache == nil {
+		cache = NewLocalCache()
+	}
+	return &CachedBackend{
+		ReportFn:      rfn,
+		cache:         cache,
+		hierarchyTree: make(hierarchyTree),
+	}
+}
+
+// NewLocalCache returns a pointer to a LocalCache with an initialised empty data structure
+func NewLocalCache() *LocalCache {
+	return &LocalCache{ds: cmap.New()}
+}
+
+// Get entries for LocalCache
+func (l LocalCache) Get(cacheKey string) (CacheValue, bool) {
+	var cv CacheValue
+	v, ok := l.ds.Get(cacheKey)
+	if !ok {
+		return cv, ok
+	}
+
+	cv = v.(CacheValue)
+	return cv, ok
+}
+
+// Set entries for LocalCache
+func (l LocalCache) Set(cacheKey string, cv CacheValue) {
+	l.ds.Set(cacheKey, cv)
+}
+
+// AuthRep provides a combination of authorizing a request and reporting metrics to 3scale
+func (cb CachedBackend) AuthRep(req AuthRepRequest, c *backend.ThreeScaleClient) (Response, error) {
+	// compute the cache key for this request
+	hierarchyKey := generateHierarchyKey(req.ServiceID, c.GetPeer())
+	appIdentifier := generateAppIdentifier(req)
+	cacheKey := generateCacheKey(hierarchyKey, appIdentifier)
+
+	var affectedMetrics backend.Metrics
+	// gather the metrics affected by this request using the hierarchy information we might have gathered
+	affectedMetrics, ok := cb.computeAffectedMetrics(hierarchyKey, req.Params.Metrics)
+
+	// instrumentation
+	mc := metricsConfig{ReportFn: cb.ReportFn, Endpoint: "AuthRep", Target: "Backend"}
+
+	if !ok {
+		// !! CACHE MISS - we don't know about this service id and host combination!!
+		// we need to build hierarchy map blanket call to authRep with extensions enabled and no usage reported
```

Comment:
Hmm, my thoughts were that since we would be reporting it in the end anyway, that the user would call `AuthRep` with that expectation, not caring when the actual reporting was done

Reply:
yes but this is referring to what we do in the implementation which the user does not know about - it is just an authorization because we have nothing to report, so just a minor correction

Reply:
ah sorry, my apologies, for some reason I assumed you were speaking of the function naming as opposed to the text here. noted and fixed